Step3 - 로또(자동)

KangSeongHyo opened the pull request:
안녕하세요~

Step3 로또(자동) 리뷰 부탁드립니다.

질문사항
- 마지막에 리팩토링을 하면서 검증처리 하는 부분을 클래스로 분리를 했는데요 이미 기존에는 테스트가 되어있어서
coverage는 동일한데 이때 분리한 클래스에 또 테스트코드를 작성하는지 궁금합니다.
-  테스트를 작성하다보니 B메서드를 테스트 하기 전에 A메서드를 호출 해야한다면 보통 어떻게 처리하는지 궁금합니다.
처음에는 일치개수처리와 수익률계산을 두개의 서비스 메서드로 분리를 했었는데 위의 고민이 있었고 이 경우에는 오히려 연관관계가 있으니 통합하는게 맞는 것 같다는 생각이 들어서 처리했고 이렇게 하는게 맞는지 궁금합니다.


감사합니다.

KangSeongHyo committed:
docs : README.md 에 FeatureList 추가

KangSeongHyo committed:
test : 입력받은 구매금액을 로또 개수로 변환 테스트 추가

KangSeongHyo committed:
feat : 입력받은 구매금액을 로또 개수로 변환 기능 추가

- 변환한 로또 개수를 Lottos에 담도록 설정

KangSeongHyo committed:
test : 구매 금액이 1000원 미만인 경우 검증 테스트 추가

KangSeongHyo committed:
feat : 구매 금액이 1000원 미만인 경우 검증 메서드 추가

- 1000원 미만인 경우 IllegalArgumentException
- ErrorMessage 클래스 추가

KangSeongHyo committed:
test : 구매 금액이 1000만원 이상인 경우 검증 테스트 추가

KangSeongHyo committed:
feat : 구매 금액이 1000만원 이상인 경우 검증 메서드 추가

- validateMoney 메서드로 분리하여 통합처리
- 1000만원 이상인 경우 IllegalArgumentException

KangSeongHyo committed:
docs : README.md FeatureList 업데이트 및 입력 금액 예외 추가

KangSeongHyo committed:
test : 구매 금액이 1000원 단위 검증 테스트 추가

KangSeongHyo committed:
feat : 구매 금액이 1000원 단위 검증 메서드 추가

KangSeongHyo committed:
docs : README.md update

KangSeongHyo committed:
test : 구간 사이의 랜덤번호 List 생성 유틸 메서드 테스트 추가

KangSeongHyo committed:
feat : 구간 사이의 랜덤번호 List 생성 유틸 메서드 추가

- low ~ max 구간에 count 만큼 랜덤 번호 생성

KangSeongHyo committed:
test : Lotto 객체 생성 테스트 추가

KangSeongHyo committed:
feat : Lotto 객체 생성 메서드 추가

- Lotto의 생성자로 RandomNumber를 받고 정렬 후 필드에 등록

KangSeongHyo committed:
test : 로또 구매개수만큼 로또게임을 진행하고 로또목록을 생성 테스트 추가

KangSeongHyo committed:
feat : 로또 구매개수만큼 로또게임을 진행하고 로또목록을 생성 메서드 추가

- Lottos는 Lotto 목록을 필드로 가지도록 추가 및 로또 등록 메서드 추가

KangSeongHyo committed:
docs : README.md complete update

KangSeongHyo committed:
test : 입력된 문자열을 구분자로 split 하고 숫자 리스트로 반환 유틸 메서드 테스트 추가

KangSeongHyo committed:
test : splitToNumberListByDelimiter에서 숫자가 아닌 경우 검증 테스트 추가

KangSeongHyo committed:
feat : 문자열을 구분자로 split해서 숫자리스트로 반환하는 메서드 추가

- 숫자가 아닌 경우 IllegalArgumentException

KangSeongHyo committed:
test : 지난 주 당첨 번호와 구매한 로또 일치 개수 계산(게임진행) 테스트 추가

KangSeongHyo committed:
feat : 지난 주 당첨 번호와 구매한 로또 일치 개수 계산(게임진행) 메서드 추가

- contains로 모든 로또에 대한 일치번호개수를 계산
- 일치번호개수를 로또목록이 가지고 업데이트 할 수 있도록 Map 필드 추가
- 결과 개수는 확인은 메서드를 통해 Map에 접근하도록 추가 및 테스트 변경

KangSeongHyo committed:
docs : README.md complete update

KangSeongHyo committed:
test : 구매금액과 총 당첨금액에 대한 수익률을 계산 테스트 추가

KangSeongHyo committed:
fix(test) : 일치 개수 테스트와 수익률 계산 테스트를 통합

- 수익률 계산 메서드를 분리하면 독립적인 기능으로 동작이 불가능하여 게임진행 안에서 처리하도록 수정

KangSeongHyo committed:
feat : playLottoGame 메서드에 수익률 계산 기능 추가

- 총 당첨금액 / 구매금액 계산 후 로또목록에 저장
- 당첨금액은 Enum으로 처리

KangSeongHyo committed:
docs : README.md complete update

KangSeongHyo committed:
docs : README.md 입력 검증 예외 상황 추가

KangSeongHyo committed:
test : 지난주 로또 번호 숫자가 1~45 사이가 아닌 경우 검증 테스트 추가

KangSeongHyo committed:
feat : 지난주 로또 번호 숫자가 1~45 사이가 아닌 경우 검증 메서드 추가

KangSeongHyo committed:
docs : README.md complete update

KangSeongHyo committed:
feat : 실행을 위한 Application 클래스 추가

KangSeongHyo committed:
feat : 금액 입력 및 로또 개수 출력에 대한 흐름 처리 추가

- 입력 오류 발생시 재시도 하도록 추가
- Client(Application) 전달을 위한 DTO 추가
- view의 경우 입력과 결과에 따른 view 클래스 추가

KangSeongHyo committed:
fix : 메서드 네이밍 변경 (playLottoByCount -> generateLottosByPlayCount)

KangSeongHyo committed:
fix : DTO 네이밍 변경 (LottoGameResponseDTO -> LottoGameDTO)

KangSeongHyo committed:
feat : 로또 생성 및 목록에 대한 흐름 처리 추가

- 로또생성 목록 결과 view 추가

KangSeongHyo committed:
fix : 로또 생성 및 목록 생성 Controller 전달 파라미터 DTO로 수정

KangSeongHyo committed:
feat : 로또게임 진행 & 결과 흐름 처리 추가

- 당첨번호 입력 처리 추가
- 오류발생시 다시 입력받도록 추가
- 당첨 통계 결과 view(일치, 수익률) 생성 추가

KangSeongHyo committed:
docs : README.md complete update

KangSeongHyo committed:
fix : 숫자가 10 이상인 경우 오류로 인식, regex + 누락 수정

KangSeongHyo committed:
docs : README.md 지난주 로또 번호 예외상황 추가 (6개가 아닌 경우)

KangSeongHyo committed:
test : 지난주 로또 번호 숫자들이 6개가 아닌 경우 검증 테스트 추가

KangSeongHyo committed:
feat : 지난주 로또 번호 숫자들이 6개가 아닌 경우 검증 메서드 추가

KangSeongHyo committed:
refactor : 공통사용 상수 클래스 분리, 로또 play 관련 검증 처리 클래스 분리

KangSeongHyo committed:
style : 컨벤션 적용

KangSeongHyo committed:
style : sonar Lint 적용

KangSeongHyo committed:
docs : README.md 지난주 로또 번호 중복인 경우 예외상황 추가

KangSeongHyo committed:
test : 지난주 로또 번호 6개의 숫자들에 중복이 있는지 검증 테스트 추가

KangSeongHyo committed:
feat : 지난주 로또 번호 6개의 숫자들에 중복이 있는지 검증 메서드 추가

KangSeongHyo committed:
style : 컨벤션 적용 (테스트 클래스)

KangSeongHyo committed:
feat : 지난주 로또 번호 입력에 대한 공백 제거 추가

KangSeongHyo committed:
docs : README.md complete update

KangSeongHyo committed:
fix(docs) : README.md Controller & Main 부분 수정 및 추가

- 재시도 처리 내용 추가
- 흐름제어 내용 추가

KangSeongHyo committed:
style : RandomNumberUtils 생성자 위치 이동

KangSeongHyo committed:
refactor(test) : 객체에 위임 하도록 Lotto 맞춘 개수 구하기 테스트 추가

KangSeongHyo committed:
refactor(feat) : 당첨번호와 로또 맞춘 개수 구하기 메서드 추가

KangSeongHyo committed:
refactor(fix) : Lotto의 NumberList getter 제거 및 연관 메서드 리팩토링

- 맞춘개수 구하기 lotto 객체에서 처리
- view 변환을 위한 Array 변환 추가

KangSeongHyo committed:
refactor(test) : convertMoneyToLottos를 대체하기 위한 구매금액으로 로또구매개수를 반환하는 테스트 추가

dhmin5693 commented on README.md:
```diff
@@ -1,9 +1,50 @@
-* [텍스트와 이미지로 살펴보는 온라인 코드 리뷰 과정](https://github.com/next-step/nextstep-docs/tree/master/codereview)
\ No newline at end of file
+
+## FeatureList
```
정리하고 시작하는 습관 매우 좋아요! 👍

dhmin5693 commented on src/main/java/lotto/model/LottoGeneratorService.java:
```diff
@@ -0,0 +1,16 @@
+
+public class LottoGeneratorService {
+
+    public Lotto generateLotto() {
```
`RandomNumberUtils` 에 너무 강하게 의존이 된 상태입니다.
의존의 방향까지는 별 문제 없습니다만 의존한 코드의 결과가 멱등하지 않아요.
특정 로직 수행 결과가 멱등하지 않으면 개발자가 코드를 정확하게 제어할 수가 없습니다.
이 때문에 예상하지 못한 문제가 발생할 수도 있고 테스트 코드도 정확하게 짜기 어려워지게 됩니다.

실제로 테스트 코드를 보면 어떤가요?
기대하는 결과가 1 ~ 45 사이의 6개 숫자임은 알 수 있지만 정확하게 어떤 값을 가지는지는 테스트로 확인하기 어렵습니다.

멱등성 문제를 가진 대표적인 예시로는 `LocalDateTime.now()`, `Random.nextInt()` 정도가 있습니다.
현재 시간을 LocalDateTime.now() 로 데이터베이스에 저장하는 코드가 있다고 해보겠습니다.
`현재 시간이 데이터베이스에 정확하게 저장된다` 라는 결과를 테스트 코드로 작성할 수 있을까요?
혹은 `2022년 5월 13일 00시 00분에 저장을 시도하면 그 시간이 정확히 기록된다`는 어떨까요?

이런 코드를 사용할 때에는 가급적이면 **외부에서 주입**받아 사용하는 형태를 추천드립니다.
사용하는 부분에서 제어할 수 있도록요.

dhmin5693 replied:
jojoldu 님의 블로그에 멱등성을 해치는 코드의 문제점을 다룬 글이 있어 첨부합니다.
https://jojoldu.tistory.com/451

KangSeongHyo replied:
이부분이 조금 어렵게 느껴졌는데요
랜덤넘버를 생성하는 부분을 Controller로 올려서 처리를 하니
클래스에 책임이 애매해지는 부분 있는 것 같은데 어떻게 하면 좋을지
이부분은 한번더 리뷰를 받아보고 싶습니다~

dhmin5693 replied:
조금 직접적으로 말씀드리자면 `LottoGeneratorService` 는 번호 생성 로직을 알지 않아도 됩니다.
만약 `NumberCreator` 라는 이름의 인터페이스에 의존하여 번호를 생성하고 있다면 어떨까요?
메인 로직에서는 랜덤값을 생성하는 로직을 그대로 사용해도 되고 테스트에서는 내멋대로 값을 주입하여 조금 더 정확한 테스트를 만들어 낼 수 있습니다.

dhmin5693 commented on src/main/java/lotto/utils/RandomNumberUtils.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    public static List<Integer> generateRandomNumberToList(int low, int max, int count) {
+        List<Integer> numberList = createNumberList(low, max);
```
메모리를 조금 더 아낄 수 있는 방법이 있지 않을까요?

KangSeongHyo replied:
Map을 통해 캐시할수 있도록 처리해보았습니다.

dhmin5693 replied:
현재 상황에서는 1 ~ 45 리스트 하나만 생성해도 되는데
아무래도 유틸성 클래스이다 보니 다른 곳에서 활용할 가능성까지 염두에 두신 점 너무 좋네요 :)

여담이지만 변수명에선 `List` 등 가급적이면 Java Collection 명을 사용하지 않으시는걸 추천드려요.
https://tecoble.techcourse.co.kr/post/2020-04-24-variable_naming

KangSeongHyo committed:
refactor(test) : convertMoneyToLottos를 대체하기 위한 구매금액으로 로또구매개수를 반환하는 검증 테스트 추가

KangSeongHyo committed:
refactor(feat) : convertMoneyToLottos를 대체하기 위한 구매금액으로 로또구매개수를 반환하는 메서드 추가

KangSeongHyo committed:
refactor : convertMoneyToLottos -> buyLottoCount 대체 및 제거(+테스트 코드)

dhmin5693 commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -0,0 +1,14 @@
+package lotto.view;
+
+public class InputView {
```
인스턴스로 생성할 필요가 없어보이는데 어떻게 생각하시나요?
클래스명이 `InputView` 이지만 사실상 `InputMessage` 의 역할을 수행하고 있는 것으로 보입니다.
`ResultView`가 `Lottos` 를 받아 출력을 수행하듯 적절한 역할을 더 할당해보시는걸 추천드립니다.

테스트가 어려운 코드 중에는 `Scanner` 등의 표준 입력도 있는데요,
`Application` 에서 해당 코드를 분리해서 여기로 책임을 이동하는건 어떤가요?
UI는 테스트 대상에서 제외니까요.

System.out.println 도 테스트 대상으로 삼기는 조금 미묘한데 `InputView`나 `ResultView`로 옮겨봅시다.
그렇게 되면 메소드 하나로 "입력해주세요" 문구와 입력을 동시에 수행할 수 있을 것 같아요.

KangSeongHyo replied:
피드백 반영하여 입력과 출력에 대한 책임을 분리해보았습니다.

dhmin5693 commented on src/main/java/lotto/model/LottoPlayService.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    public void playLottoGame(Lottos lottos, List<Integer> winningNumberList) {
+        LottoGameValidateUtils.validateWinningNumberList(winningNumberList);
```
로또 번호는 `Lottos` 에서 관리하게 만드셨더라구요.
당첨 번호도 **하나의 객체**에 책임을 할당할 수 있지 않을까요?
그럼 이 validation 로직도 이동시킬 수 있을 것 같아요.

KangSeongHyo replied:
``WinningLotto`` 객체로 분리해 보았습니다.

dhmin5693 commented on src/main/java/lotto/constant/LottoMatchNumber.java:
```diff
@@ -0,0 +1,31 @@
+        this.winningAmount = winningAmount;
+    }
+
+    public static Set<LottoMatchNumber> allMatchNumber() {
```
LottoMatchNumber.values() 로도 대체할 수 있는데 혹시 EnumSet으로 만들어 둔 특별한 이유가 있을까요?

KangSeongHyo replied:
이부분 확인하여 수정했습니다.

dhmin5693 commented on src/main/java/lotto/vo/Lottos.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    public void setResultProfitRate(double resultProfitRate) {
+        this.resultProfitRate = resultProfitRate;
```
setter는 제거해주세요.
`Lottos` 가 책임을 너무 많이 가진게 근본적인 원인같아 보입니다.
적절하게 분배해보는건 어떨까요?

KangSeongHyo committed:
refactor(feat) : Lottos 객체지향 할수있도록 책임 분리

- addLotto 대신 LottoList를 파라미터로 하는 Lottos 생성자 생성
- generateLottosByCount -> generateLottoByCount 대체

dhmin5693 commented on src/main/java/lotto/controller/LottoGameController.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    public LottoGameDTO inputMoney() {
+        return new LottoGameDTO(null, inputView.inputMoneyView(), false);
```
메소드 명으로 동작을 유추하기가 힘들어보여요.
실제로 하는 동작은 dto에 메시지를 넣어서 생성하는건데 메소드명은 금액 입력이거든요.

금액 자체의 역할을 별도로 분리해봐도 좋겠네요.

dhmin5693 commented on src/main/java/lotto/dto/LottoGameDTO.java:
```diff
@@ -0,0 +1,28 @@
+
+import lotto.vo.Lottos;
+
+public class LottoGameDTO {
```
dto로 데이터를 이동시키는건 각 레이어 간 직접 연결을 줄여주는 아주 좋은 수단입니다 👍 
하지만 dto 하나를 너무 많은 곳에서 사용하고 있어서 불필요한 필드가 많이 생기는 것 같아요.

실제로 `new LottoGameDTO` 로 인스턴스를 생성하는 곳은 8군데인데 그 중 5군데에서 `Lottos` 를 null로 넣고 있습니다.

KangSeongHyo replied:
이부분은 객체의 책임을 분리하기 시작하니 DTO가 사용이 두군데로 줄어서
 Application으로 예외 전달에 필요하니 생성자를 새로 추가해서 처리해보았는데
 혹시 Application과 Controller를 더 나누는게 좋은지 질문드립니다.

dhmin5693 replied:
`Application` 과 `Controller` 는 지금으로도 충분한 듯 합니다.
피드백을 잘 반영해주셔서 dto 의존도가 매우 많이 낮아졌네요 👍 

별도의 `Service` 가 없어지면서 `LottoGameDTO` 자체도 최종적으론 삭제할 수 있을 것 같긴 한데,
지금으로선 충분하니 고민 정도만 한번 해보셔도 좋을 것 같아요!

dhmin5693 reviewed (changes requested):
안녕하세요 성효님.
미션 진행해주시느라 굉장히 수고 많으셨어요.

몇 가지 개선점을 남겨보았습니다.
남겨드린 개선점의 대부분은 `객체의 책임` 과 관련되어 있어요.
TDD 미션은 객체를 더욱 객체답게 설계하는 것을 목표로 하고 있거든요.

코드 스타일을 보면 자바와 스프링 프레임워크에 친숙하신 것처럼 보입니다.
MVC 구조로 변경해보자는 의미는 스프링에서 주로 사용되는 레이어드 아키텍처 기반의 코드 구조로 작성하기보다는 각 객체의 역할을 최소한으로 쪼개보자는 의미에 더 가깝습니다.

기분좋은 주말에 너무 많은 과제를 남겨드린 것 같아 마음이 무겁네요.
미션 진행하다가 어렵다고 생각되는 점 있으시다면 바로 DM 주세요.
언제든지 도와드리겠습니다.
저희는 그걸 위해 이 일을 하고 있으니까요 :)

KangSeongHyo committed:
refactor : 로또개수 변환과 로또 생성에 대한 결과를 통합하여 처리하도록 변경

KangSeongHyo committed:
refactor : 콘솔입력에 대한 책임을 Appication에서 분리

KangSeongHyo committed:
refactor : 출력에 대한 책임을 View로 이동

KangSeongHyo committed:
refactor : 불필요하게 파라미터 길어지지 않도록 LottoGameController static 변수로 변경

KangSeongHyo committed:
refactor : 구매금액 관련 검증 책임들을 LottoStore 클래스로 이관

KangSeongHyo committed:
refactor : 로또개수에 대한 책임을 LottoPaper로 분리

KangSeongHyo committed:
refactor(test) : 로또를 생성에 대한 책임을 LottoGenerator로 분리하기 위한 생성 테스트 추가

KangSeongHyo committed:
refactor(feat) : LottoGenerator에서 Lotto를 생성하도록 정적 메서드 추가

KangSeongHyo committed:
refactor : Lottos, Lotto 패키지 이동 (vo -> model)

KangSeongHyo committed:
refactor(test) : 당첨번호에 대한 검증을 WinningLotto로 책임 분리하기 위한 테스트 이관

KangSeongHyo committed:
refactor(feat) : 당첨번호에 대한 검증을 WinningLotto로 책임을 생성자로 이관 추가

KangSeongHyo committed:
refactor(fix) : 아키텍쳐 수정 View에서 출력까지 담당하도록 변경

KangSeongHyo committed:
refactor(fix) : 로또 생성에 대한 부분을 각 객체별 연결을 통해 할수 있도록 수정

KangSeongHyo committed:
refactor : 로또생성에대한 객체 책임 이관에 따른 LottoGeneratorService 제거

KangSeongHyo committed:
refactor : 로또 개수 생성에 대한 책임을 LottoStore로 이관, 불필요 메서드 제거

KangSeongHyo committed:
refactor(test) : 로또들에 대한 비교를 WinningLotto로 이관하기 위한 테스트 추가

KangSeongHyo committed:
refactor(feat) : 로또 등수에 대한 Enum 추가 및 등수 계산 추가

KangSeongHyo committed:
refactor(feat) : 로또 결과(등수,총상금)에 대한 책임을 LottoGameResult로 분리

KangSeongHyo committed:
refactor : LottoPlayService 레이어 제거

KangSeongHyo committed:
refactor : 검증 처리에 대한 유틸 클래스 각각 책임 클래스로 분리 완료로 제거

KangSeongHyo committed:
refactor : 로또 수익률 계산 Lottos로 이관

KangSeongHyo committed:
refactor : 각각 로또결과에 대한 분리한 객체들로 결과 처리 및 view 생성하도록 변경

KangSeongHyo committed:
refactor : LottoGameDTO 에서 view 필드 제거

KangSeongHyo committed:
refactor : 리팩토링으로 불필요한 테스트 및 코드 정리

KangSeongHyo committed:
refactor : LottoGameDTO 오류 발생에 대한 생성자 추가

KangSeongHyo committed:
refactor : shuffle 대상인 NumberList cacheMap 추가

KangSeongHyo committed:
refactor : 메서드 네이밍 리팩토링

KangSeongHyo committed:
refactor : Lottos 에 add 메서드를 제거하고 생성자로 초기화하도록 변경

KangSeongHyo committed:
refactor : 유틸클래스 생성할 수 없도록 방어로직 추가

KangSeongHyo committed:
refactor : 불필요 클래스 제거

KangSeongHyo committed:
refactor : Rank에 NONE 제외한 메서드를 가져올수 있도록 변경

KangSeongHyo committed:
style : 컨벤션 적용

KangSeongHyo committed:
style : 메서드 위치 변경

KangSeongHyo committed:
style : InputStringUtils의 메서드명 변경

KangSeongHyo committed:
refactor : LottoGameMessage 에서 LottoInputMessage 로 클래스명 변경

KangSeongHyo committed:
refactor : LottoGameController를 LottoGame으로 변경 및 정적메서드 형태로 변경

dhmin5693 commented on src/main/java/lotto/utils/RandomNumberUtils.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    private static List<Integer> createRandomNumberList(int count, List<Integer> shuffleNumberList) {
+        List<Integer> randomNumberList = new ArrayList<>();
```
이 부분은 특별한 로직 없이 한 줄로 끝내도 괜찮습니다.
```
return new ArrayList<>(shuffleNumberList);
```

KangSeongHyo replied:
아 이부분은 count 개수 만큼만 추가하고 있습니다.

dhmin5693 commented on src/main/java/lotto/vo/Lotto.java:
```diff
@@ -0,0 +1,26 @@
+                .count());
+    }
+
+    public int[] numberListToArray() {
```
public 메소드에 대한 테스트는 반드시 작성해주세요!

dhmin5693 commented on src/test/java/lotto/vo/LottoTest.java:
```diff
@@ -0,0 +1,23 @@
+    @Test
+    void matchNumberCount() {
+        Lotto lotto = new Lotto(Arrays.asList(1, 3, 15, 25, 36, 45));
+        assertAll(
```
한 번에 검증하여 테스트 대상을 정확히 구분하시는 모습 매우 좋습니다 👍

dhmin5693 reviewed (approved):
하루만에 늘어난 커밋의 개수를 보고 성효님이 얼마나 미션을 열심히 진행해주셨는지
그 열기가 잘 느껴지네요 👍

코드를 전체적으로 한번 뒤엎는 과정을 거치면서 이번 미션의 목적을 충분히 달성하셨다고 생각합니다.
간단한 코멘트 몇 개 남겼는데요, 큰 건 아니니 잠깐 확인 정도만 해보시길 바랍니다.

다음 미션 진행해주세요.
정말 고생 많으셨습니다.
